src/dotnet/commands/dotnet-tool/restore/ToolRestoreCommand.cs

```diff
@@ -96,117 +97,111 @@ public override int Execute()
+                        verbosity: _verbosity);
+
+                if (!ManifestCommandMatchesActualInPackage(package.CommandNames, toolPackage.Commands))
                 {
-                    toolPackageExceptions.Add(package.PackageId, e);
+                    return ToolRestoreResult.Failure(
+                        string.Format(LocalizableStrings.CommandsMismatch,
+                            JoinBySpaceWithQuote(package.CommandNames.Select(c => c.Value.ToString())),
+                            package.PackageId,
+                            JoinBySpaceWithQuote(toolPackage.Commands.Select(c => c.Name.ToString()))));
                 }
-            }
-
-            EnsureNoCommandNameCollision(dictionary);
 
-            _localToolsResolverCache.Save(dictionary, _nugetGlobalPackagesFolder);
-
-            return PrintConclusionAndReturn(dictionary.Count() > 0, toolPackageExceptions, errorMessages, successMessages);
+                return ToolRestoreResult.Success(
+                    saveToCache: toolPackage.Commands.Select(command => (
+                        new RestoredCommandIdentifier(
+                            toolPackage.Id,
+                            toolPackage.Version,
+                            NuGetFramework.Parse(targetFramework),
+                            Constants.AnyRid,
+                            command.Name),
+                        command)).ToArray(),
+                    message: string.Format(
+                        LocalizableStrings.RestoreSuccessful,
+                        package.PackageId,
+                        package.Version.ToNormalizedString(),
+                        string.Join(" ", package.CommandNames)));
+            }
+            catch (ToolPackageException e)
+            {
+                return ToolRestoreResult.Failure(package.PackageId, e);
+            }
         }
 
-        private int PrintConclusionAndReturn(
-            bool anySuccess,
-            Dictionary<PackageId, ToolPackageException> toolPackageExceptions,
-            List<string> errorMessages,
-            List<string> successMessages)
+        private int PrintConclusionAndReturn(ToolRestoreResult[] toolRestoreResults)
         {
-            if (toolPackageExceptions.Any() || errorMessages.Any())
+            if (toolRestoreResults.Any(r => !r.IsSuccess))
             {
                 _reporter.WriteLine(Environment.NewLine);
                 _errorReporter.WriteLine(string.Join(
-                                        Environment.NewLine,
-                                        CreateErrorMessage(toolPackageExceptions).Concat(errorMessages)).Red());
+                    Environment.NewLine,
+                    toolRestoreResults.Where(r => !r.IsSuccess).Select(r => r.Message)).Red());
 
                 _reporter.WriteLine(Environment.NewLine);
 
-                _reporter.WriteLine(string.Join(Environment.NewLine, successMessages));
+                _reporter.WriteLine(
```

Comment:
Consider that this will write an extra blank line if there are no success messages.